[Doc] Translate runner design (#710)
open-mmlab/mmengine: docs/en/design/runner.md

# Runner

Deep learning algorithms usually share similar pipelines for training, validation and testing.

> Therefore, MMengine designed `Runner` to simplify the construction of this pipeline.

Reviewer: ```suggestion
Therefore, MMengine designed `Runner` to simplify the construction of these pipelines.
```

In most cases, users can use our default `Runner` directly.
If you find it not feasible to implement your ideas, you can also modify it or customize your own runner.

> Before introducing the design of `Runner`, let's walk through some examples to better understand why should we use runner.

Reviewer 2: ```suggestion
Before introducing the design of `Runner`, let's walk through some examples to better understand why we should use runner.
```

These are all achieved by runner and various practical modules in MMEngine.

![Runner](https://user-images.githubusercontent.com/12907710/184577204-3ea033bd-91dd-4da8-b4ac-22763d7d6c7d.png)

The `Runner` in MMEngine contains various modules required for training, testing and validation, as well as loop controllers(`Loop`) and [Hook](../tutorials/hook.md), as shown in the figure above.
It provides 3 APIs for users: `train`, `val` and `test`, each correspond to a specific `Loop`.
You can use `Runner` either by providing a config file, or by providing manually constructed modules.
Once activated, the `Runner` will automatically setup the runtime environment, build/compose your modules, execute the loop iterations in `Loop` and call registered hooks during iterations.

The execution order of `Runner` is as follows:

![runner_flow](https://user-images.githubusercontent.com/12907710/184577118-b8f30521-0dba-4b94-a78f-8682459650a5.png)

A feature of `Runner` is that it will always lazily initialize modules managed by itself.
To be specific, `Runner` won't build every module on initialization, and it won't build a module until it is needed in current `Loop`.

> Therefore, you only need to provide a limited set of configs/modules if you are running only one of `train`, `val` or `test` pipeline.

Reviewer: ```suggestion
Therefore, if you are running only one of the `train`, `val`, or `test` pipelines, you only need to provide the relevant configs/modules.
```

## Loop

In MMEngine, we abstract the execution process of the task into `Loop`, based on the observation that most deep learning tasks can be summarized as a model iterating over datasets.
We provide 4 built-in loops in MMEngine:

- EpochBasedTrainLoop
- IterBasedTrainLoop
- ValLoop
- TestLoop

![Loop](https://user-images.githubusercontent.com/12907710/184577588-d74e16dd-15c7-4f73-9857-61c56c29057b.png)

The built-in runner and loops are capable of most deep learning tasks, but surely not all.
Some tasks need extra modifications and refactorizations.
Therefore, we make it possible for users to customize their own pipelines for model training, validation and testing.

> You can write your own pipeline by subclassing [BaseLoop](mmengine.runner.BaseLoop), which needs 2 arguments for initialization: 1) `runner` the Runner instance, and 2) `loader` the dataloader used in this loop.

Reviewer: ```suggestion
You can write your own pipeline by subclassing [BaseLoop](mmengine.runner.BaseLoop), which needs 2 arguments for initialization: 1) `runner` the Runner instance, and 2) `dataloader` the dataloader used in this loop.
```
